File: calculator.ex
defmodule Calculator do
    @moduledoc """
    Provide mathematical calculation from user input
    """
    @operators ~w(+ - * / ^ %)
    
    @doc """
    Calculate x and y according to the user choice operant
    """
    def calculate(x, operator, y) when is_number(x) and is_number(y) and operator in @operators,
    do: execute(x, operator, y)
    # handles invalid operator
    def calculate(_, operator, _) when operator not in @operators, do: {:error, "Invalid operator"}
    def calculate(_, _, _), do: {:error, "Invalid arguments! Ex. `calculate(5, +, 5)`"}
     # executes the calculation
    defp execute(x, "+", y), do: x + y
    defp execute(x, "-", y), do: x - y
    defp execute(x, "*", y), do: x * y
    defp execute(_, "/", 0), do: {:error, "Division by zero"}
    defp execute(x, "/", y), do: x / y
    defp execute(x, "^", y), do: :math.pow(x, y)
    defp execute(x, "%", y), do: rem(x, y)
end

defmodule Main do
    @doc """
    Get user input for the numbers to be calculated
    """
    def get_user_input do
        number = IO.gets("Enter a number: ")
        if number not in [""] do
            elem(Float.parse(number), 0)
        else
            nil
        end 
    end

    @doc """
    Get the user input operator for our calculation
    """
    def get_operator do
        IO.gets("Enter an operator sign [+ - / * % ^]: ")
        |> String.trim()
    end

    def run do
        IO.puts("\t==============================\n")
        IO.puts("\t Welcome to ElixirCalculator! \n")
        IO.puts("\t==============================\n\n")
        x        = get_user_input()
        operator = get_operator()
        y        = get_user_input()

        case Calculator.calculate(x, operator, y) do
            {:error, message} -> IO.puts "Error: #{message}"
            result -> IO.puts "Result: #{x} #{operator} #{y} = #{result}"
        end
    end
end

File: calculator.go
/*
 * Simple app calculator written in Golang
 * Author: Chris Dedman
 */
package main

import (
	"errors"
	"fmt"
	"log"
	"math"
	"strconv"
)

// Get user input for the numbers to be calculated
// Return the user number
func get_user_input() float64 {
	var input string

	fmt.Printf("Enter a number: ")
	fmt.Scan(&input)

	number, err := strconv.ParseFloat(input, 64)
	if err != nil {
		fmt.Printf("Invalid input. Please enter a valid number.\n")
		return get_user_input()
	}

	return number
}

// Get the user input operator for our calculation
// Return the operator sign
func get_operator() string {
	var operator string
	fmt.Print("Enter an operator sign [+ - / * % ^]: ")
	fmt.Scan(&operator)

	return operator
}

// Calculate x and y according to the user choice operant
// Return the value calculated
func calculator(x float64, operant string, y float64) (float64, error) {
	switch operant {
	case "+":
		return x + y, nil
	case "*":
		return x * y, nil
	case "-":
		return x - y, nil
	case "/":
		if y == 0 {
			return y, errors.New("'Division by zero'")
		}
		return x / y, nil
	case "%":
		return math.Mod(x, y), nil
	case "^":
		return math.Pow(x, y), nil
	default:
		return x, errors.New("Invalid operator")
	}
}

func main() {
	log.SetPrefix("Error: ")
	log.SetFlags(0)

	fmt.Printf("\t==========================\n")
	fmt.Printf("\t Welcome to GoCalculator! \n")
	fmt.Printf("\t==========================\n\n")

	first_number := get_user_input()
	operant := get_operator()
	second_number := get_user_input()

	result, err := calculator(first_number, operant, second_number)
	if err != nil {
		log.Fatal(err)
	}
	fmt.Printf("\nResult: %v %v %v = %v\n", first_number, operant, second_number, result)
}
